File: Java/Library.java
import java.util.ArrayList;
import java.util.Scanner;

public class Library {
   private ArrayList<Book> books = new ArrayList<Book>();
   // Let's assume username and password are not actually stored here, but on another file.
   private static String username = "test";
   private static String password = "password1";

   public Library() {
      books.add(new Book("The Great Gatsby", "F. Scott Fitzgerald", 3));
      books.add(new Book("To Kill a Mockingbird", "Harper Lee", 2));
      books.add(new Book("1984", "George Orwell", 1));
      books.add(new Book("Fahrenheit 451", "Ray Bradbury", 1));
   }

   public void displayAvailableBooks() {
      System.out.println("Available books:");
      for (Book book : books) {
         if (book.getNumCopies() > 0) {
            book.display();
         }
      }
   }

   public void checkout(String title) {
      for (Book book : books) {
         // CWE-486: Classes not compared solely by name
         if (!book.getClass().equals(Book.class)) {
            System.out.println("Sorry, an error occurred");
            return;
         }
         if (book.getTitle().equalsIgnoreCase(title)) {
            if (book.getNumCopies() > 0) {
               book.setNumCopies(book.getNumCopies() - 1);
               System.out.println("You have checked out \"" + title + "\". Enjoy!");
            } else {
               System.out.println("Sorry, there are no copies of \"" + title + "\" available.");
            }
            return;
         }
      }
      System.out.println("Sorry, we do not have \"" + title + "\" in our library.");
   }

   public void returnBook(String title) {
      for (Book book : books) {
         // CWE-481: Book titled is being compared and not assigned
         if (book.getTitle().equalsIgnoreCase(title)) {
            book.setNumCopies(book.getNumCopies() + 1);
            System.out.println("You have returned \"" + title + "\". Thank you!");
            return;
         }
      }
      System.out.println("Sorry, we do not have \"" + title + "\" in our library.");
   }

   public static void main(String[] args) {
      Library library = new Library();
      Scanner scanner = new Scanner(System.in);
      String choice = "";
      String title;
      StringBuffer text1 = new StringBuffer("");
      StringBuffer text2 = new StringBuffer("");

      System.out.println("Hello! Welcome to our library. We hope you enjoy your visit!");
      System.out.print("Please enter your username: ");
      text1.append(scanner.nextLine());
      System.out.print("Please enter your password: ");
      text2.append(scanner.nextLine());

      // Let's also assume here that instead of using an if/else, we can also use a try/catch statement in order to
      // query a SQL database using the supplied username-password to retrieve info and "log in", with a 
      // catch statement to supply the user with an error if the login fails.
      //
      // CWE-209: In the "else" statement (instead of a catch), we do not generate a message/error that 
      // supplies the user with sensitive information, like information about the SQL query logic or 
      // the database like names of tables/columns that could lead to SQL injection.
      if (text1.toString().equals(username) && text2.toString().equals(password))
      {
         // CWE-201: Sensitive information such as username and password are given little chance to be accessed by other actors before being removed.
         // CWE-212: Sensitive information such as username and password are properly removed before too much is done with them available.
         text1.delete(0, text1.length());
         text2.delete(0, text2.length());

         while (!choice.equalsIgnoreCase("4"))
         {
            System.out.print("What would you like to do?\n\t[1] Display available books\n\t[2] Check out a book\n\t[3] Return a book\n\t[4] Quit\nChoice:  ");
            choice = scanner.nextLine();
            switch (choice.toLowerCase()) {
               case "1":
                  library.displayAvailableBooks();
                  break; // CWE-484: Break statements are utilized in each case for the switch statement
               case "2":
                  System.out.println("Enter the title of the book you want to check out: ");
                  title = scanner.nextLine();
                  library.checkout(title);
                  break;
               case "3":
                  System.out.println("Enter the title of the book you are returning: ");
                  title = scanner.nextLine();
                  library.returnBook(title);
                  break;
               default:
                  // CWE-478: A default case is implemented for the multiple condition expression
                  System.out.println("Invalid choice. Please try again.");
                  break;
            }
         }
      }
      else
      {
         System.out.println("Username or password invalid.");
      }

      System.out.println("We hope you enjoyed your trip to our library!");
      scanner.close();
   }
}

File: C++/Library.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <iostream>
#include <ctime>
#include "Book.h"

// CWE-480: Proper usage of operators can be observed throughout the code, including arithmetic and logical operators.
// CWE-457: The books vector is initialized to a default value before it is used
// CWE-482: Proper use of comparing vs assigning
std::vector<Book> availableBooks = {
    Book("The Great Gatsby", "F. Scott Fitzgerald", 3),
    Book("To Kill a Mockingbird", "Harper Lee", 2),
    Book("1984", "George Orwell", 2),
    Book("Fahrenheit 451", "Ray Bradbury", 1)
};

std::vector<Book> checkedOutBooks = {};

void displayBooks()
{
    for (auto& book : availableBooks)
    {
        // CWE-595: Proper comparing of object values instead of references
        std::cout << "\"" << book.getTitle() << "\" by " << book.getAuthor() << " (" << book.getQuantity() << " available)" << std::endl;
    }
}

void checkOutBook(std::string title)
{
    // CWE-480: Proper usage of operators can be observed throughout the code, particularly arithmetic and logical operators.
    // CWE-483: Proper block delimitation can also be observed throughout the code, particularly with if statements.
    for (auto& book : availableBooks)
    {   
        // CWE-482: Proper use of comparing vs assigning
        // CWE-595: Proper comparing of object values instead of references
        if (book.getTitle() == title)
        {
            if (book.getQuantity() > 0)
            {
                book.setQuantity(book.getQuantity() - 1);
                // Add book to checkedOutBooks
                checkedOutBooks.push_back(Book(book.getTitle(), book.getAuthor(), 1));
                std::cout << "You have checked out \"" << book.getTitle() << "\" by " << book.getAuthor() << "." << std::endl;
            }
            else
            {
                std::cout << "Sorry, there are no copies of \"" << book.getTitle() << "\" available." << std::endl;
            }
            return;
        }
    }
    std::cout << "Sorry, we do not have \"" + title + "\" in our library." << std::endl;
}

void returnBook(std::string title)
{
    for (auto& book : availableBooks)
    {   
        // CWE-482: Proper use of comparing vs assigning
        // CWE-595: Proper comparing of object values instead of references
        if (book.getTitle() == title)
        {
            // CWE-120: Check to make sure the value provided in setQuantity is within the size of its (integer) buffer.
            // CWE-125: Checking within the size of a buffer doubles as checking within the bounds of it.
            // CWE-482: Proper use of comparing vs assigning
            // CWE-595: Proper comparing of object values instead of references
            if (book.getQuantity() - INT32_MAX != 0)
            {
                // CWE-595: Proper comparing of object values instead of references
                book.setQuantity(book.getQuantity() + 1);
            }
            // CWE-482: Proper use of comparing vs assigning
            bool flag = false;
            // CWE-595: Proper comparing of object values instead of references
            for (int i = 0; i < ((int) checkedOutBooks.size()); i++) {
                // CWE-482: Proper use of comparing vs assigning
                if (checkedOutBooks[i].getTitle().compare(title) == 0) {
                    checkedOutBooks.erase(checkedOutBooks.begin() + i);
                    // CWE-482: Proper use of comparing vs assigning
                    flag = true;
                }
            }
            if (flag) {
                book.setQuantity(book.getQuantity() + 1);
                std::cout << "You have returned \"" << book.getTitle() << "\". Thank you!" << std::endl;
            } else {
                std::cout << "Sorry, you don't have \"" << title << "\" checked out." << std::endl;
            }
            return;
        }
    }
    std::cout << "Book not found." << std::endl;
}

void printReceipt()
{
    // CWE-482: Proper use of comparing vs assigning
    std::time_t t = std::time(nullptr);
    std::tm* now = std::localtime(&t);
    std::tm returnDate = *now;
    returnDate.tm_mday += 14;
    std::mktime(&returnDate);

    // Open output file stream
    std::ofstream outFile("receipt.txt");

    // CWE-397: Throws a specific error (not just exception)
    // CWE-595: Proper comparing of object values instead of references
    if (!(outFile.is_open())) {
        std::runtime_error("Could not open file");
    }

    // Write receipt header
    outFile << "RECEIPT:\n\n";

    // Write book titles and return date
    for (auto& book : checkedOutBooks) {
        // CWE-595: Proper comparing of object values instead of references
        outFile << book.getTitle() << "\nRETURN BY: "
                << returnDate.tm_year + 1900 << "/"
                << returnDate.tm_mon + 1 << "/"
                << returnDate.tm_mday << "\n\n";
    }

    // Close output file stream
    // CWE-910: Proper use of expired file descriptor
    outFile.close();
}


int main()
{
    // CWE-482: Proper use of comparing vs assigning
    int choice = 0;
    bool usingLibrary = true;
    std::string input;

    std::cout << "Hello! Welcome to our library. We hope you enjoy your visit!\n";

    while (usingLibrary)
    {
        std::cout << "What would you like to do?\n\t[1] Display available books\n\t[2] Check out a book\n\t[3] Return a book\n\t[4] Print receipt\n\t[5] Quit\nChoice:  ";
        std::cin >> input;
        try
        {
            // CWE-482: Proper use of comparing vs assigning
            // CWE-704: Proper use of converting or casting
            choice = stoi(input);
        }
        // CWE-210: If an error occurs, there are no specifics given on why an error occurred other than "invalid choice" at the end of the switch.
        catch(const std::exception& e)
        {
            // CWE-482: Proper use of comparing vs assigning
            choice = 0;
        }

        switch (choice)
        {
            case 1:
                displayBooks();
                break;
            case 2:
                std::cout << "Enter the title of the book: ";
                std::cin.ignore();
                getline(std::cin, input);
                checkOutBook(input);
                break;
            case 3:
                std::cout << "Enter the title of the book: ";
                std::cin.ignore();
                getline(std::cin, input);
                returnBook(input);
                break;
            case 4:
                // CWE-248: Catches potential exception
                try {  
                    printReceipt();
                    std::cout << "Your receipt has been successfully printed!\n";
                }
                // CWE-396: Uses specific error (not just exception) 
                // CWE-460: Program is in a safe state once exception is caught
                catch (const std::runtime_error& e) {
                    std::cout << "Error: " << e.what() << "\n";
                }
                break;
            case 5:
                // CWE-482: Proper use of comparing vs assigning
                usingLibrary = false;
                break;
            default:
                std::cout << "Invalid choice. Please try again." << std::endl;
        }
    }
    return 0;
}
